e2e/backend_service_update_test.go

```diff
@@ -0,0 +1,200 @@
+	return tcprID, nil
+}
+
+func buildBackendServiceWithLBScheme(graphBuilder *rgraph.Builder, name string, lbScheme string) (*cloud.ResourceID, error) {
+	bsID := backendservice.ID(testFlags.project, meta.GlobalKey(resourceName(name)))
+
+	bsMutResource := backendservice.NewMutableBackendService(testFlags.project, bsID.Key)
+	bsMutResource.Access(func(x *compute.BackendService) {
+		x.LoadBalancingScheme = lbScheme
+		x.Protocol = "TCP"
+		x.PortName = "http"
+		x.SessionAffinity = "NONE"
+		x.Port = 80
+		x.TimeoutSec = 30
+		x.ConnectionDraining = &compute.ConnectionDraining{}
+	})
+	bsResource, err := bsMutResource.Freeze()
+	if err != nil {
+		return nil, err
+	}
+
+	bsBuilder := backendservice.NewBuilder(bsID)
+	bsBuilder.SetOwnership(rnode.OwnershipManaged)
+	bsBuilder.SetState(rnode.NodeExists)
+	bsBuilder.SetResource(bsResource)
+
+	graphBuilder.Add(bsBuilder)
+	return bsID, nil
+}
+
+func checkTCPRoute(t *testing.T, ctx context.Context, cloud cloud.Cloud, tcprID *cloud.ResourceID) {
+	t.Helper()
+	t.Log("---- Check TCP Route ---- ")
+	_, err := cloud.TcpRoutes().Get(ctx, tcprID.Key)
+	if err != nil {
+		t.Fatalf("cloud.TCPRoutes().Get(_, %s) = %v, want nil", tcprID.Key, err)
+	}
+}
+
+func checkBackendService(t *testing.T, ctx context.Context, cloud cloud.Cloud, bsID *cloud.ResourceID, bsPort int) {
+	t.Helper()
+	t.Log("---- Check BackendService ----")
+	gotBS, err := cloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("cloud.HealthChecks().Get(_, %s) = %v, want nil", bsID.Key, err)
+	}
+
+	if gotBS.Port != int64(bsPort) {
+		t.Errorf("BackendService port mismatch, got: %d want %d", gotBS.Port, bsPort)
+	}
+}
+func TestBackendServiceUpdate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+
+	bsName := "bs-e2e"
+	bsID, err := buildBackendServiceWithLBScheme(graphBuilder, bsName, "INTERNAL_SELF_MANAGED")
+	if err != nil {
+		t.Fatalf("buildSimpleBackendService(_, %s, _) = %v, want nil", resourceName(bsName), err)
+	}
+
+	serviceName := fmt.Sprintf("https://compute.googleapis.com/v1/projects/%s/global/backendServices/%s", testFlags.project, resourceName(bsName))
+	tcpr1, err := buildTcpRoute(graphBuilder, "route1", serviceName)
+	t.Logf("tcpr1 = %s", pretty.Sprint(tcpr1))
+	if err != nil {
+		t.Fatalf("buildTcpRoute(_, route1, %s) = %v, want nil", serviceName, err)
+	}
+
+	tcpr2, err := buildTcpRoute(graphBuilder, "route2", serviceName)
+	t.Logf("tcpr2 = %s", pretty.Sprint(tcpr1))
+	if err != nil {
+		t.Fatalf("buildTcpRoute(_, route2, %s) = %v, want nil", serviceName, err)
+	}
+
+	graph, err := graphBuilder.Build()
+	if err != nil {
+		t.Fatalf("graphBuilder.Build() = %v, want nil", err)
+	}
+
+	result, err := plan.Do(ctx, theCloud, graph)
+	if err != nil {
+		t.Fatalf("plan.Do(_, _, _) = %v, want nil", err)
+	}
+
+	ex, err := exec.NewSerialExecutor(result.Actions)
+	if err != nil {
+		t.Logf("exec.NewSerialExecutor(_, _) err: %v", err)
+		return
```

Comment:
Wouldn't `t.Logf + return` cause test to exit without failing first? I think `t.Fatalf` would be better

Reply:
Done.